### ISSUE TITLE
--incremental Watch Mode

### ISSUE THREAD

**Issue by zachleat** [11ty/eleventy#108]
Only rebuild what changed (if templates). If includes, data, et al—rebuild everything.

**Comment by zachleat** [comment]
The really difficult part about this task is that doing it correctly would require a detailed dependency map of the content, specifically around what uses what collections.

**Comment by insanetisane** [comment]
I believe the pug cli watch option does this. Even if included css file changes it only rebuilds templates that reference it. 
It’s very smart. Not sure how it’s done tho.

**Comment by zachleat** [comment]
Proxies and Sorted Directed Acyclic Graphs, y’all.

https://www.npmjs.com/package/toposort
https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Proxy

**Comment by zellwk** [comment]
Just adding another thought: I wonder how would incremental watch mode work with a task runner like Gulp and browser sync?

**Comment by jakearchibald** [comment]
If/when this happens, it'd be great if the optimisations were available outside of watch mode.

TypeScript does this via `.tsbuildinfo` files.

This means that users that are using different watchers would benefit from the incremental build too.

**Comment by Ryuno-Ki** [comment]
Hm, some people referred different existing implementations. Could we gather their source code here so to lay out a plan?
(If we get lucky, we find shared libraries among them!)

**Comment by zachleat** [comment]
Did a really rough start of this here https://github.com/11ty/eleventy/commit/6e294bd7160d44b400810bab84273336009c99db

**Comment by zachleat** [comment]
After working on https://www.zachleat.com/web/own-my-tweets/ I’m also thinking a nice feature might be some smarts about pagination data too, specifically adding entries into a huge paginated array would only write files for the new/changed entries.

**Comment by zachleat** [comment]
@matthewp had an interesting idea about reading files in the output directory to compare contents to avoid re-writing and save on rsync tasks https://twitter.com/matthewcp/status/1174013502215852032

**Comment by zachleat** [comment]
I did a fair bit of work to experiment playing around with @matthewp’s idea to compare output before writing is in the `incremental-file-hash-cache` branch but it wasn’t quite good enough. Will require more work but might get scrapped. https://github.com/11ty/eleventy/tree/incremental-file-hash-cache

Hashing the content seemed to be fast enough (using `farmhash` https://www.npmjs.com/package/farmhash) but writing an extra file for the cache seemed to negate the gains here.

Some super quick numbers (farmhash versus node’s built-in crypto module)

![image](https://user-images.githubusercontent.com/39355/65332137-b9728680-db83-11e9-970d-28da30368f2b.png)

Yeah, not writing any files did provide a small performance improvement for large projects (but at the cost of up-front hashing for the first run to create the cache file)

I dunno, just not quite sold here on this yet.

**Comment by matthewp** [comment]
@zachleat Thanks for giving it a try! For clarity, I wasn't expecting the comparison to speed up builds, I was expecting it to speed up `rsync`.

Out of curiosity, was this hashing implementation a separate hash file for every page, or did you concat these some how into 1 hashed file so you only need to do 1 read.

From an implementation simplicity perspective you could skip the hashing and just read in the current version of the file. This would indeed be slower, but how much slower?

**Comment by piperhaywood** [comment]
Another thought about this:
If `--incremental` or something similar is too difficult, a different approach that might help is to exclude files via the config file instead of `.eleventyignore`. This way, we can use the preexisting [`--config` flag](https://www.11ty.io/docs/usage/#--config-to-change-the-config-file-name) to swap configs when developing locally.

My specific use case:
We’re pulling in data from an API on each build. Right now there are only 20 pages on the whole site, but the build time is taking anywhere from 9 to 50 seconds each time due to the API. Ideally we wouldn’t hit that API every time, it’s only really necessary when the site is deployed to Netlify. If I could exclude that particular file within an Eleventy config file such as `eleventy.dev.js`, then I could run `npx @11ty/eleventy --serve --config=eleventy.dev.js` for much faster development.

**Comment by dceddia** [comment]
This feature would be awesome to have. I was thinking about how to do the dependency mapping and how it seems like each template language would need to be able to do its own map.

Poking around Nunjucks docs I found the `load` event ([here](https://mozilla.github.io/nunjucks/api.html#load-event)). It sounds potentially useful! They say:

> The 'load' event gets emitted whenever a Loader retrieves the source of a template. It can be listened to in order to determine template dependencies at runtime.

Still gotta build up that DAG I guess, but this might be a step in the right direction, for Nunjucks anyway.

**Comment by bep** [comment]
> I wasn't expecting the comparison to speed up builds, I was expecting it to speed up rsync.

rsync has a --checksum flag.

On an added note, I'm glad to see that I'm not the only having an hard time wrapping my head around "partial builds" :-)

**Comment by zachleat** [comment]
I’m dividing this work into smaller issues and orchestrating on a GitHub project here: https://github.com/orgs/11ty/projects/3

**Comment by Lohn** [comment]
Just wondering... And If we use some command to to build a specific file? Not the builder find this file, but we explicitly pointing it. 
Something like:

eleventy --build-incremental _blog/2020/some_nice_title_here.md

I asking about this, because when you update one post/note/photo, you (our your CMS) know what was changed. Maybe is easier (and less resourceful) this way.

**Comment by Ryuno-Ki** [comment]
I doubt, that this would result in a nice experience.
Imagine, that you have a year index page for all blog pages published in 2020.
This one wouldn't be updated, which could lead in confusion.

In technical terms: There's a dependency tree, which would flag other sites as „dirty”, i.e. in need of recomputation.

**Comment by Lohn** [comment]
I understand. But anyway the incremental build won't identify only the file changed?
So, it needs to rebuild the index page too. I was wondering about a way to avoid the "hash search" of all the files to make the incremental build less  time/resource consuming.

**Comment by AleksandrHovhannisyan** [comment]
@Ryuno-Ki Not necessarily. Jekyll does this very intelligently by only rebuilding a file under one of two conditions:

1. If the file itself was modified. For example, if you modify a blog post or a page, it will get rebuilt.
2. If the file is included by other files/pages, or if it's a layout for other pages, then all those dependencies will get rebuilt.

Right now, I'm seeing build times of only 4-5 seconds with Jekyll 4 whenever a single file changes. With 11ty, that increases to 20 seconds because everything is getting rebuilt every single time. Eventually, if I save a file frequently enough, it just runs out of memory and crashes.

**Comment by uxdiscovery** [comment]
Really looking forward to this epic feature. I'm working with an i18n project which eventually generates 5000+ pages. That takes 15 mins to finish a build

**Comment by Takazudo** [comment]
Isn't eleventy interested in incremental build?
The fact that eleventy does not have this feature is the huge disadvantage against other SSG solutions, I guess.

**Comment by Ryuno-Ki** [comment]
https://github.com/orgs/11ty/projects/3 lists the feature as in-progress.

**Comment by Takazudo** [comment]
@Ryuno-Ki Oh I understood. Thank you for the guidance!

As a person who develops web sites and web applications for client works, I want you to know that we can't choose eleventy for the big projects because of this.

**Comment by Ryuno-Ki** [comment]
What a bummer!
Perhaps you can find another great tool at https://jamstack.org/generators/ that suits your requirements better.

**Comment by AleksandrHovhannisyan** [comment]
Actually, now that I've revisited working on migrating my site to 11ty, I'm curious: Isn't this how the `--incremental` flag already behaves? I've noticed that it doesn't actually rebuild the entire site, which is a good thing. I must've not been using it before.

**Comment by pbuzdin** [comment]
https://www.11ty.dev/docs/usage/incremental/

**Comment by j-f1** [comment]
A neat addition to this would be to have `--serve` mode only build/copy files that are requested. This would mean that changing a file would immediately tell Browsersync to reload, which would then tell Eleventy to trigger a build of everything needed by that file (optionally skipping if none of the dependencies have changed).

**Comment by zachleat** [comment]
@j-f1 related to https://github.com/11ty/eleventy/issues/108#issuecomment-1003446526 https://github.com/11ty/eleventy/issues/2456

**Comment by rr923** [comment]
@zachleat Can I use this feature for production build? I noticed that the document says: 
`Incremental builds perform a partial build operating only on files that have changed to improve build times when doing local development.`

**Comment by zachleat** [comment]
@rr923 Work on higher fidelity incremental builds is happening and I will likely add more issues to reflect more of the nuances but for this to offer improvements to production builds on your CI server we’ll need cold start incremental first:

https://www.11ty.dev/docs/usage/incremental/#cold-start

**Comment by zachleat** [comment]
Docs updated: https://www.11ty.dev/docs/usage/incremental/

As of 2.0.0-canary.21 we’re now shipping:

* Incremental builds for layout file changes
* Incremental builds for collections consume/publish
* an [API for dependency mapping](https://www.11ty.dev/docs/languages/custom/#registering-dependencies) that is used for incremental builds (for custom templates), already in use by [the WebC plugin](https://www.11ty.dev/docs/languages/webc/)
* [For-free passthrough copy](https://www.11ty.dev/docs/copy/#passthrough-during-serve)

Further issues have been filed for future improvements:

* [Cold start](https://github.com/11ty/eleventy/issues/984)
* [Data files](https://github.com/11ty/eleventy/issues/2706)
* [Automated collections dependencies](https://github.com/11ty/eleventy/issues/1615)

This is finally in good enough shape that I’m ready to say that Eleventy now has incremental builds support. I’m going to milestone `--incremental` as complete in 2.0 🏆
